### ISSUE TITLE
[Feature Request] Allow for SafeModules with in_keys and out_keys of depth > 1

### ISSUE THREAD

**Issue by ajlangley** [pytorch/rl#927]
## Motivation

The main motivation for making this change is for easier compatibility with multi-agent environments. For example, the Pettingzoo parallel API returns observations, and accepts actions, as dictionaries with agent names as keys. Therefore, in order to manipulate these return values, networks need to both operate on and return values that may be more than 1 level deep in a dictionary. For example, the batch returned by a collector may look like

```
{
   ...
    "observation": {
        "Agent1": A tensor,
        ...
        "AgentN": A tensor
    }
}
```

Then, each agent's actor module could have as it's `in_key`, `'('observation', 'agent_id')`. 

The main advantage here would be that multiple agents' modules could all be wrapped in a `SafeSequential` module, which would allow us to write code for multi-agent algorithms with minimal modifications relative to their single-agent counterparts.

I understand that `SafeModule` performs a check as to whether all keys are strings. I would love to know why this is necessary, since the `TensorDictModule` itself provides the desired functionality already.

## Solution

Allow `SafeModule`s to use tuples of keys in their `in_keys` and `out_keys`.

## Alternatives

I have considered creating a flat version of the input dictionary by prepending/appending the agent name to every key. This works fine but will require extra wrappers/transformations to/for existing environments.

- [ x ] I have checked that there is no similar issue in the repo (**required**)

**Comment by vmoens** [comment]
Let me know if I got this wrong but I think this is supported

```python
>>> from torchrl.modules import SafeModule
>>> from tensordict import TensorDict
>>> import torch
>>> 
>>> module = SafeModule(torch.nn.Linear(3, 4), in_keys=[("some", "key", "I", "want")], out_keys=[("some", "key", "I", "do", "not", "have")])
>>> data = TensorDict({("some", "key", "I", "want"): torch.randn(3)}, [])
>>> module(data)
>>> print(data)
TensorDict(
    fields={
        some: TensorDict(
            fields={
                key: TensorDict(
                    fields={
                        I: TensorDict(
                            fields={
                                do: TensorDict(
                                    fields={
                                        not: TensorDict(
                                            fields={
                                                have: Tensor(shape=torch.Size([4]), device=cpu, dtype=torch.float32, is_shared=False)},
                                            batch_size=torch.Size([]),
                                            device=None,
                                            is_shared=False)},
                                    batch_size=torch.Size([]),
                                    device=None,
                                    is_shared=False),
                                want: Tensor(shape=torch.Size([3]), device=cpu, dtype=torch.float32, is_shared=False)},
                            batch_size=torch.Size([]),
                            device=None,
                            is_shared=False)},
                    batch_size=torch.Size([]),
                    device=None,
                    is_shared=False)},
            batch_size=torch.Size([]),
            device=None,
            is_shared=False)},
    batch_size=torch.Size([]),
    device=None,
    is_shared=False)

```

**Comment by ajlangley** [comment]
You are correct, so I was mistaken. I realized from trying your code that it is specifically `ValueModule`s which don't support this:

**Comment by vmoens** [comment]
Will fix!

**Comment by ajlangley** [comment]
Hi, I remembered why I thought the issue was with `SafeModule`s. It appears that the `ProbabilisticActor` class is the one that will not accept depth >1 keys. Any chance this functionality could be added?

**Comment by vmoens** [comment]
Should be fixed by https://github.com/pytorch-labs/tensordict/pull/239
you can now have nested input and output keys in `ProbabilisticTensorDictModule`
